### PR TITLE
Add a Dynamo deepdive to documentation

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122305

This supersedes the previous `Guards Overview" as a more comprehensive
approach to most of the main topics within Dynamo.

In the future, we could add specific sections for each of the topics
discussed here.

### CODE REVIEW

File: docs/source/torch.compiler_dynamo_deepdive.rst [hunk 1]
```diff
@@ -0,0 +1,873 @@
...
+```OutputGraph`` <https://github.com/pytorch/pytorch/blob/69f112d5867f785a3a090a0c6d6644ae047033ac/torch/_dynamo/output_graph.py#L221-L230>`__
+object. The ``OutputGraph`` object is `bound to an
+``InstructionTranslator``
+object <https://github.com/pytorch/pytorch/blob/69f112d5867f785a3a090a0c6d6644ae047033ac/torch/_dynamo/symbolic_convert.py#L2060-L2071>`__
+and it tracks all the data necessary to create the FX graph which will
+be returned by Dynamo.
+
+All the inputs and intermediary elements of the FX graph are
+``fx.Node``\ s. In Dynamo, ``fx.Node``\ s are wrapped in
+``fx.Proxy``\ s, which record operations done on them in the graph. We
```
Comment (msaroufim) [inline docs/source/torch.compiler_dynamo_deepdive.rst:342]: nit: can you elaborate what record operations done on them means?

File: docs/source/torch.compiler_dynamo_deepdive.rst [hunk 2]
```diff
@@ -0,0 +1,873 @@
...
+```wrap_fx_proxy`` <https://github.com/pytorch/pytorch/blob/fb80f05ee2e1cba17892980701bfd5dbce58349f/torch/_dynamo/variables/builder.py#L1311>`__.
+
+A graph stores operations on tensors… and operations on symbolic
+integers. We will discuss symbolic integers later on, but first we will
+discuss how Dynamo addresses a rather important correctness issue.
+
+Making Dynamo Sound: Guards
+---------------------------
+
+At this point, we have a way to trace linear programs. And for that, we
```
Comment (msaroufim) [inline docs/source/torch.compiler_dynamo_deepdive.rst:355]: what's a linear program?

File: docs/source/torch.compiler_dynamo_deepdive.rst [hunk 3]
```diff
@@ -0,0 +1,873 @@
...
+   def fn(a, b):
+       return a * len(b)
+
+   fn(torch.arange(10), "Hello")
+
+with ``TORCH_LOGS=guards`` prints (among other guards)
+
+.. code:: python!
+
+   ___check_type_id(L['b'], 94334122025024)
```
Comment (msaroufim) [inline docs/source/torch.compiler_dynamo_deepdive.rst:395]: where does the 9433.. number come from?

File: docs/source/torch.compiler_dynamo_deepdive.rst [hunk 4]
```diff
@@ -0,0 +1,873 @@
...
+
+   fn(torch.randn(8), ["Hi", "Hello"])
+
+generates the following guards
+
+.. code:: python!
+
+   ___check_type_id(L['l'], 94439025877664)
+   len(L['l']) == 2
+   hasattr(L['x'], '_dynamo_dynamic_indices') == False
```
Comment (msaroufim) [inline docs/source/torch.compiler_dynamo_deepdive.rst:469]: worth elaborating in a few more lines what this is

File: docs/source/torch.compiler_dynamo_deepdive.rst [hunk 5]
```diff
@@ -0,0 +1,873 @@
...
+   l_a_ (concrete): (8, 3)
+   l_b_: (s0, 3)
+   l_b_ (concrete): (8, 3)
+   mul: (s0, 3)
+   mul (concrete): (8, 3)
+   mul_1: (s0, 3)
+   mul_1 (concrete): (8, 3)
+
+where we can see that the first dimension of the two tensor args is
+dynamic.
```
Comment (msaroufim) [inline docs/source/torch.compiler_dynamo_deepdive.rst:563]: dynamic because there's symbolic values s0

**Review by msaroufim — APPROVED** [review]
really enjoyed reading this, thank you for writing it

File: docs/source/torch.compiler_dynamo_deepdive.rst (hunk 4 above)
Comment (lezcano) [reply docs/source/torch.compiler_dynamo_deepdive.rst:469]: Just removed that line to avoid distracting the reader.

**Pull request closed without merging**